### PR TITLE
Support inplace move of VarBase

### PR DESCRIPTION
**Pull request by Aurelius84**
### PR types

New features
### PR changes

APIs
### Describe

Support inplace move of VarBase

### CODE REVIEW

**Review by zhwesky2010 — DISMISSED** [review]
LGTM

File: paddle/fluid/imperative/layer.cc [hunk 1]
```diff
@@ -311,6 +312,42 @@ void VarBase::CopyFrom(const VarBase& src, const bool blocking) {
...
+  const framework::Tensor* src_tensor = nullptr;
+  if (Var().IsType<framework::LoDTensor>()) {
+    src_tensor = &Var().Get<framework::LoDTensor>();
+  } else {
+    src_tensor = &Var().Get<framework::SelectedRows>().value();
+  }
+
+  framework::Tensor dst_tensor;
+  dst_tensor.Resize(src_tensor->dims());
+  dst_tensor.mutable_data(dst_place, src_tensor->type());
```
Comment (liym27) [inline paddle/fluid/imperative/layer.cc:337]: Does the difference between `dst_tensor` and `src_tensor`  **only** is **the place**? 
If yes,  
1. `dst_tensor` need to share the `inplace_version_counter_` with `src_tensor` to ensure the correctness of gradient calculation
Please add code like:
```c++
dst_tensor->ShareInplaceVersionCounterWith(src_tensor);
```

2. Here only set the same `dims` and `type` as `src_tensor`, does it ensure that the other attributes of `src_tensor` and `dst_tensor` are consistent? such as `offset_`, `layout_`?

Comment (Aurelius84) [reply paddle/fluid/imperative/layer.cc:337]: Thanks to find this, I will fix this.

Comment (Aurelius84) [reply paddle/fluid/imperative/layer.cc:337]: use `src_tensor->ShareBufferWith(dst_tensor)` to reset `holder_` and `offset_`. Other information of src_tensor is still keep unchange.

**Pull request closed without merging**